internal/bucket/bandwidth/monitor.go

```diff
@@ -19,52 +19,110 @@ package bandwidth
 
 import (
 	"context"
+	"math"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/minio/madmin-go"
+	"golang.org/x/time/rate"
 )
 
-// throttleBandwidth gets the throttle for bucket with the configured value
-func (m *Monitor) throttleBandwidth(ctx context.Context, bucket string, bandwidthBytesPerSecond int64, clusterBandwidth int64) *throttle {
+const (
+	throttleInternal = 250 * time.Millisecond
+)
+
+type throttle struct {
+	*rate.Limiter
+	NodeCount           uint64
+	NodeBandwidthPerSec int64
+	freeBytes           int64 // unused bytes in the interval
+	bytesPerInterval    int64 // bytes allocated for the interval
+	goGenerate          int64 // Flag to track if generate routine should be running. 0 == stopped
+}
+
+// ReduceAvailableBandwidth reduces extra bandwidth that was allocated for user from
+// available bytes
+func (m *Monitor) ReduceAvailableBandwidth(bucket string, bytes int64) {
 	m.lock.Lock()
 	defer m.lock.Unlock()
-	throttle, ok := m.bucketThrottle[bucket]
-	if !ok {
-		throttle = newThrottle(ctx, bandwidthBytesPerSecond, clusterBandwidth)
-		m.bucketThrottle[bucket] = throttle
-		return throttle
+	if t, ok := m.bucketThrottle[bucket]; ok {
+		m.generateBandwidthForBucket(bucket, t.freeBytes-bytes)
 	}
-	throttle.SetBandwidth(bandwidthBytesPerSecond, clusterBandwidth)
-	return throttle
 }
 
-// Monitor implements the monitoring for bandwidth measurements.
-type Monitor struct {
-	lock sync.Mutex // lock for all updates
+func (m *Monitor) updateMeasurement(bucket string, bytes uint64) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m, ok := m.activeBuckets[bucket]; ok {
+		m.incrementBytes(bytes)
+	}
+}
+func (m *Monitor) generateBandwidthForBucket(bucket string, bytes int64) {
+	t := m.bucketThrottle[bucket]
+	t.freeBytes = bytes
+	bw := math.Max(0, float64(t.freeBytes))
+	var newrate rate.Limit
+	if bw > 0 {
+		newrate = rate.Every(throttleInternal / time.Duration(bw))
+	}
+	t.SetLimitAt(time.Now(), newrate)
+}
 
-	activeBuckets map[string]*bucketMeasurement // Buckets with objects in flight
+// GenerateBandwidthForBucket generates available bandwidth for bucket for this interval
+func (m *Monitor) GenerateBandwidthForBucket(bucket string) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if t, ok := m.bucketThrottle[bucket]; ok {
+		m.generateBandwidthForBucket(bucket, t.bytesPerInterval)
+	}
+}
 
-	bucketMovingAvgTicker *time.Ticker // Ticker for calculating moving averages
+// generateBandwidth periodically allocates new bandwidth to use
+func (m *Monitor) generateBandwidth(ctx context.Context) {
+	for {
+		select {
+		case <-m.generateTicker.C:
+			for b, t := range m.bucketThrottle {
+				if atomic.LoadInt64(&t.freeBytes) == atomic.LoadInt64(&t.bytesPerInterval) {
+					// No bandwidth consumption stop the routine.
+					atomic.StoreInt64(&t.goGenerate, 0)
+					return
```

Comment:
If we return here we would not be allocating new bandwidth for buckets that follow this bucket with no consumption.

Reply:
this is refactored a bit to handle the data race - does use a return because this is at the bucket specific throttle